### ISSUE TITLE
Label Parsers for Messages

### ISSUE THREAD

**Issue by cdamus** [eclipsesource/papyrus-seqd#16]
## Summary

The lightweight sequence diagram needs GMF *Label Parsers* for presentation and direct-editing of message names in the diagram.

### R1. Message numbering
The user must have the option to
* **R1.1** show or hide message numbering in the diagram
* **R1.2** when showing message numbers, use either
  * **R1.2.1** simple numbers:  1, 2, 3, ...
  * **R1.2.2** hierarchical numbers:  1, 1.1, 1.2, 1.2.1, 1.3, 2, ...

This numbering is apart from any other content of a message label, as described in following sections.  Moreover, it is not accessible in the direct editor; it is strictly computed from the sequencing of messages in the interaction model.  Message numbers are presented before (to the left of in left-to-right locales) the message label as specified by UML.

Hierarchical numbering conforms to the *sequence expression* specification for *Communication Diagrams* in §17.9.1.3 of the [UML 2.5.1 specification](https://www.omg.org/spec/UML/2.5.1/).  This includes the `a`, `b`, etc. names for concurrent messages in a parallel combined fragment.

### R2. Message Signatures
If the message references a behavioral feature as its *signature*, then a textual representation of that feature should be shown as the label.  This will typically be an `Operation` of the receiving class (actual behavioral feature) or a `Signal` (implied behavioral feature) for which the receiving class has a `Reception` (that being the actual feature).  Or, in the case of a component *port* receiving the message, an `Operation` of some provided `Interface`.

*Note that, per §17.4.3.1 of the [UML 2.5.1 specification](https://www.omg.org/spec/UML/2.5.1/), a message that has a signature must have the same name as that signature.  Accordingly, the name will be fixed and not editable for any message that has a signature.*

### R3. Direct Editing
Support for direct editing of the message label must conform to the label syntax prescribed in §17.4.4.1 of the [UML 2.5.1 specification](https://www.omg.org/spec/UML/2.5.1/).  Three input schemes are offered for the user to choose from (in the diagram preferences):
* **R3.1** structured label:  as per the specification cited above, with syntax highlighting and content-assist for selection of *signature* names, operation parameter names, and signal attribute names from the contextual model based on the type of the `Lifeline`'s represented `ConnectableElement`.  Conformance to the UML standard syntax is enforced by the in-line editor
* **R3.2** free-form label:  the user can type any text.  No syntax highlighting, content assist, or validation is provided
* **R3.3** hybrid:  the user can type any text, with content-assist to select the name of the *signature* element.  No validation is provided, nor syntax highlighting

As indicated in the description of **R1**, above,
* **R3.4** the sequence number (if any) is not presented in the *Direct Editor* for editing by the user

**Comment by acampesino** [comment]
You may want to have a look at the gerrit contribution we did for neon, using xtext.

[https://git.eclipse.org/r/#/c/85194/](https://git.eclipse.org/r/#/c/85194/)

Synch and Asynch messages are implemented, but no implementation for create or return messages.

**Comment by planger** [comment]
We need to double-check whether the implementation is aligned with the UML specification (see chapter 17.4.4 page 579ff).

**Comment by cdamus** [comment]
That is exactly the section on which basis the current implementation was created.

Are we looking for more test coverage?  What is the deliverable expected?

**Comment by planger** [comment]
Sorry, I don't have a specific scenario that is clearly non-compliant... I just wanted to take a note that we should double-check and make sure it is compliant. I'll try to do some testing with the current implementation wrt the UML spec.

**Comment by planger** [comment]
I think we can close this issue, can't we? Message labels seem to be implemented in a compliant manner now and update correctly, as far as we tested them.

**Comment by cdamus** [comment]
> I think we can close this issue, can't we? Message labels seem to be implemented in a compliant manner now and update correctly, as far as we tested them.

Yes, we have display of the message labels, but have we integrated direct editing?  With support for the signature and arguments?

**Comment by planger** [comment]
No, but there are a couple of other issues on this topic already: #273 
But we can move this issue to Milestone 6, where direct editing is due to keep the information.
Thanks!
